base/compiler/typeinfer.jl

```diff
@@ -918,6 +918,33 @@ function typeinf_code(interp::AbstractInterpreter, method::Method, @nospecialize
     return code, rt
 end
 
+"""
+    typeinf_ircode(
+        interp::AbstractInterpreter,
+        method::Method,
+        atype,
+        sparams::SimpleVector,
+    ) -> (ir::Union{IRCode,Nothing}, returntype::Type)
+
+Infer a `method` and return an optimized `IRCode` with inferred `returntype` on success.
+"""
+function typeinf_ircode(interp::AbstractInterpreter, method::Method, @nospecialize(atype), sparams::SimpleVector)
+    mi = specialize_method(method, atype, sparams)::MethodInstance
+    ccall(:jl_typeinf_begin, Cvoid, ())
+    result = InferenceResult(mi)
+    frame = InferenceState(result, :no, interp)
+    if !(frame !== nothing && typeinf(interp, frame) && frame.inferred)
+        ccall(:jl_typeinf_end, Cvoid, ())
+        return nothing, Any
+    end
+    opt_params = OptimizationParams(interp)
+    opt = OptimizationState(frame, opt_params, interp)
+    ir = run_passes(opt.src, opt, result)
+    ccall(:jl_typeinf_end, Cvoid, ())
+    rt = widenconst(ignorelimited(result.result))
+    return ir, rt
+end
```

Comment:
```suggestion
function typeinf_ircode(interp::AbstractInterpreter, method::Method, @nospecialize(atype), sparams::SimpleVector)
    frame = typeinf_frame(interp, method, atype, sparams, false)
    frame === nothing && return nothing, Any
    opt_params = OptimizationParams(interp)
    opt = OptimizationState(frame, opt_params, interp)
    ir = run_passes(opt.src, opt, result)
    rt = widenconst(ignorelimited(result.result))
    return ir, rt
end
```

Reply:
I just want to double-check, but is it OK to run the optimizer outside the typeinf lock?

cc @vtjnash @pchintalapudi